code.js

```diff
@@ -1,3 +1,53 @@
 function tsp_hk(distance_matrix) {
-    return -1;
+    tsp_mems = [];
+    if(distance_matrix.length < 1) {
+        return 0;
+    } else {
+    for(let i = 0; i < distance_matrix.length; i++){
+        temp = tsp_HeldKarp(distance_matrix, i)
+    }
+    return temp
+    }
+}
+
+function tsp_HeldKarp(distance_matrix, start) {
+    if(tsp_mems[JSON.stringify(distance_matrix) + start] === undefined) {
+        if(distance_matrix.length <= 1) {
+            return 0;
+        } else if(distance_matrix.length == 2) {
+            return distance_matrix[0][1];
+        } else {
+            let min = Infinity;
+            let minI = -1;
+            matrix = matrixShrink(distance_matrix, start)
+            for(let i = 0; i < matrix.length; i++) {
+                temp = tsp_HeldKarp(matrix, i)
+                if(temp < min) {
+                    min = temp;
+                    minI = i + 1;
+                }
+            }
+            tsp_mems[JSON.stringify(distance_matrix) + start] = min + distance_matrix[start][minI];
+            return tsp_mems[JSON.stringify(distance_matrix) + start];
+        }
+    } else {
+            return tsp_mems[JSON.stringify(distance_matrix) + start];
+        }
+}
+
+function matrixShrink(matrix, rowCol) { // Removes a full list from an adjacency matrix
```

Comment:
This might be easier if you just keep track of the list of cities to still visit.